File: examples/companion-packages/zig-zon-app/src/match_engine.zig
//! Pure match-engine helpers used by the `pawmatch match` command.

const std = @import("std");
const variants = @import("variants.zig");

pub const MatchPreferences = struct {
    has_kids: bool = false,
    quiet_home: bool = false,
    active: bool = false,
    first_time: bool = false,
    multiple_pets: bool = false,
    small_home: bool = false,

    pub fn isEmpty(self: MatchPreferences) bool {
        return !(self.has_kids or self.quiet_home or self.active or
            self.first_time or self.multiple_pets or self.small_home);
    }

    pub fn isActive(self: MatchPreferences, factor_flag: []const u8) bool {
        if (std.mem.eql(u8, factor_flag, "has-kids")) return self.has_kids;
        if (std.mem.eql(u8, factor_flag, "quiet-home")) return self.quiet_home;
        if (std.mem.eql(u8, factor_flag, "active")) return self.active;
        if (std.mem.eql(u8, factor_flag, "first-time")) return self.first_time;
        if (std.mem.eql(u8, factor_flag, "multiple-pets")) return self.multiple_pets;
        if (std.mem.eql(u8, factor_flag, "small-home")) return self.small_home;
        return false;
    }
};

pub const MatchFactor = struct {
    flag: []const u8,
    tags: []const []const u8,
};

pub const match_factors = [_]MatchFactor{
    .{ .flag = "has-kids", .tags = &.{ "good-with-kids", "gentle" } },
    .{ .flag = "quiet-home", .tags = &.{ "mellow", "calm", "solo", "lap-cat" } },
    .{ .flag = "active", .tags = &.{ "high-energy", "playful" } },
    .{ .flag = "first-time", .tags = &.{ "gentle", "calm", "low-energy" } },
    .{ .flag = "multiple-pets", .tags = &.{"social"} },
    .{ .flag = "small-home", .tags = &.{ "lap-cat", "solo", "low-energy" } },
};

pub const popularity_tags = [_][]const u8{ "social", "good-with-kids", "calm", "mellow", "gentle" };

pub fn factorsForDepth(depth: variants.MatchDepth) []const MatchFactor {
    const take: usize = switch (depth) {
        .short => 2,
        .standard => 4,
        .thorough => 6,
    };
    const bound = @min(take, match_factors.len);
    return match_factors[0..bound];
}

pub fn containsTag(tags: []const []const u8, target: []const u8) bool {
    for (tags) |t| {
        if (std.mem.eql(u8, t, target)) return true;
    }
    return false;
}

pub fn countMatchingTags(tags: []const []const u8, targets: []const []const u8) usize {
    var n: usize = 0;
    for (tags) |t| {
        if (containsTag(targets, t)) n += 1;
    }
    return n;
}

File: examples/companion-packages/rust-cargo-app/src/match_engine.rs
//! Pure match-engine helpers used by the `pawmatch match` command.

use std::collections::HashSet;

use crate::variants::MatchDepth;

#[derive(Debug, Clone, Default)]
pub struct MatchPreferences {
    pub has_kids: bool,
    pub quiet_home: bool,
    pub active: bool,
    pub first_time: bool,
    pub multiple_pets: bool,
    pub small_home: bool,
}

impl MatchPreferences {
    pub fn is_empty(&self) -> bool {
        !(self.has_kids
            || self.quiet_home
            || self.active
            || self.first_time
            || self.multiple_pets
            || self.small_home)
    }

    pub fn active_factors(&self) -> HashSet<&'static str> {
        let mut out: HashSet<&'static str> = HashSet::new();
        if self.has_kids {
            out.insert("has-kids");
        }
        if self.quiet_home {
            out.insert("quiet-home");
        }
        if self.active {
            out.insert("active");
        }
        if self.first_time {
            out.insert("first-time");
        }
        if self.multiple_pets {
            out.insert("multiple-pets");
        }
        if self.small_home {
            out.insert("small-home");
        }
        out
    }
}

#[derive(Debug, Clone, Copy)]
pub struct MatchFactor {
    pub flag: &'static str,
    pub tags: &'static [&'static str],
}

pub const MATCH_FACTORS: &[MatchFactor] = &[
    MatchFactor {
        flag: "has-kids",
        tags: &["good-with-kids", "gentle"],
    },
    MatchFactor {
        flag: "quiet-home",
        tags: &["mellow", "calm", "solo", "lap-cat"],
    },
    MatchFactor {
        flag: "active",
        tags: &["high-energy", "playful"],
    },
    MatchFactor {
        flag: "first-time",
        tags: &["gentle", "calm", "low-energy"],
    },
    MatchFactor {
        flag: "multiple-pets",
        tags: &["social"],
    },
    MatchFactor {
        flag: "small-home",
        tags: &["lap-cat", "solo", "low-energy"],
    },
];

pub const POPULARITY_TAGS: &[&str] = &["social", "good-with-kids", "calm", "mellow", "gentle"];

pub fn factors_for_depth(depth: MatchDepth) -> &'static [MatchFactor] {
    let take = match depth {
        MatchDepth::Short => 2,
        MatchDepth::Standard => 4,
        MatchDepth::Thorough => 6,
    };
    let bound = take.min(MATCH_FACTORS.len());
    &MATCH_FACTORS[..bound]
}

pub fn count_tag_matches(tags: &[&str], target: &HashSet<&str>) -> usize {
    tags.iter().filter(|t| target.contains(*t)).count()
}
